Libraries/acdriver/Sources/CompileActionImageSet.cpp

```diff
@@ -0,0 +1,498 @@
+/**
+ Copyright (c) 2015-present, Facebook, Inc.
+ All rights reserved.
+
+ This source code is licensed under the BSD-style license found in the
+ LICENSE file in the root directory of this source tree. An additional grant
+ of patent rights can be found in the PATENTS file in the same directory.
+ */
+
+#include <acdriver/CompileActionImageSet.h>
+#include <acdriver/CompileAction.h>
+#include <acdriver/Result.h>
+#include <xcassets/Asset/ImageSet.h>
+#include <xcassets/Slot/Idiom.h>
+#include <libutil/Filesystem.h>
+#include <libutil/FSUtil.h>
+#include <bom/bom.h>
+#include <car/Reader.h>
+#include <car/Writer.h>
+#include <car/Facet.h>
+#include <car/Rendition.h>
+#include <bom/bom_format.h>
+
+#include <algorithm>
+#include <fstream>
+#include <map>
+#include <string>
+
+#include <png.h>
+
+using acdriver::CompileAction;
+using acdriver::CompileOutput;
+using acdriver::Options;
+using acdriver::Result;
+
+using libutil::FSUtil;
+
+static bool readPNGFile(const char *filename, std::vector<unsigned char> &pixels, size_t *width_out, size_t *height_out, size_t *channels_out, Result *result)
+{
+    char header[8];
+    png_struct *png_struct_ptr;
+    png_info *info_struct_ptr;
+
+    int number_of_passes;
+    png_uint_32 row_bytes;
+    png_byte **row_pointers = NULL;
+
+    FILE *fp = fopen(filename, "rb");
```

Comment:
Can you use the `Filesystem` abstraction for this, so it can be unit tested? You'd need to load the PNG from memory.